request: Replace deprecated request module with net.request.

manavmehta opened the pull request:
**Any background context you want to provide?**
Fixes: #886 

**You have tested this PR on:**
  - [X] macOS Catalina 10.15.5
  - [ ] Windows
  - [ ] Linux/Ubuntu


**Checklist**

- [x] domain-util
- [x] reconnect-util
- [x] linuxupdater
- [x] check add-new-server
- [x] check UA on test server

We may think of a more appropriate and suitable name of the new added request.ts

manavmehta committed:
Replace deprecated request module with electron's native net.request

The request module has been deprecated by electron and this commit replaces the same with net.request.
The net api is a main process api and thus the functions have been moved for convention.

Fixes: #886

Co-authored-by: Anders Kaseorg <anders@zulip.com>
Signed-off-by: Anders Kaseorg <anders@zulip.com>

andersk commented on app/renderer/js/utils/domain-util.ts:
```diff
@@ -1,17 +1,14 @@
 import * as Messages from '../../../resources/messages';
 
-const {app, dialog} = remote;
+const {app, dialog, net, session} = remote;
```
Unfortunately I don’t think `electron.net` can be used safely via `remote`; see electron/electron#12545. I think we’ll need to build our own IPC interface and use it in the main process. (Which is just as well, because `remote` is also [discouraged](https://github.com/electron/electron/blob/master/docs/tutorial/security.md#15-disable-the-remote-module) and we should aim to eliminate it eventually.)

abhigyank replied:
So it would be better to move all `electron.net` requests to the main process via a IPCRenderer/Main interface call?

Following this up, we could have a separate module in the main process that handles the replacement of remote calls in the renderer process? Solving #915 .

manavmehta replied:
Okay, I was facing the `end` problem too. Seems plausible. Thanks @andersk .
Also, 
> IPC primitives (`ipcRenderer` and `ipcMain`) will be unaffected. This only affects the `remote` module, which is a convenience layer wrapped on top of those IPC primitives.
> 
> The recommended migration path for apps is to replace usage of the remote module with explicit use of the IPC primitives, for example [`ipcRenderer.invoke`](https://electronjs.org/docs/api/ipc-renderer#ipcrendererinvokechannel-args) / [`ipcMain.handle`](https://electronjs.org/docs/api/ipc-main#ipcmainhandlechannel-listener).
>

Is it safe to say that the the `primitives` are safer and better, performance-wise, to use and hence to migrate to?
Are there any cases where they lag behind `remote`?

manavmehta commented:
Also, for the record, [`remote` is now available as a package](https://github.com/electron/electron/issues/21408#issuecomment-648398795) (on it's way to be moved out from electron core)

andersk commented:
Let’s move discussion about `remote` to #915.

(But to answer your questions in short: yes, we should eliminate `remote`. Yes, we should eliminate it completely. No, there are no cases of it that should not be eliminated. It’s being moved from core to a package because it should *not* be used, not because it should be used. In case you haven’t read this page, which I linked from #915 and is also linked from the Electron bug and the `@electron/remote` documentation, see [Electron’s ‘remote’ module considered harmful](https://medium.com/@nornagon/electrons-remote-module-considered-harmful-70d69500f31). I think we’re all on the same page here. Good? Good.)

abhigyank commented:
The original functionality remains intact for me after the migration. Further, now we don't need certificate-util, I tested adding certificates to Chrome store and using `certutil`on Linux, and server gets added without a hitch.
All in all LGTM. 
One minor issue, in case of certificate error on server addition, we show two dialog box errors - one from electron.net and one we create. But I think that's a bug in how we handle certificate errors, could be a follow-up PR.

manavmehta commented:
Going with [this approach](https://chat.zulip.org/#narrow/stream/16-desktop/topic/remove.20ignoreCerts), I think it will handle two dialog boxes.

abhigyank commented on app/main/request.ts:
```diff
@@ -0,0 +1,148 @@
+							// Following check handles both the cases
+							icon: realm_icon.startsWith('/') ? realm_uri + realm_icon : realm_icon,
+							url: realm_uri,
+							alias: escape(realm_name)
```
Need to import `import escape from 'escape-html';` .
Currently the PR uses built-in `escape`, which's behavior is different from what we have in `domain-util.ts` .

abhigyank commented:
Let's keep getProxy for now? I feel it completes the the utility of a file like `proxy-util` though unused. Could be wrong here.
Removing request-util is correct.

andersk commented:
Disagree. There’s no point in keeping around code that won’t be used with `electron.net`. If it burns, set it on fire.

abhigyank commented:
I agree it won't be used with electron.net, however I disagree that keeping it could be worse. I disagree also that we should fire to anything that burns, houses, cars everything burns, I don't suppose we should set fire to them. Just because anything burns, we shouldn't burn them.

andersk commented:
Uh…unused code is not a house or a car. Unused code is a readability complication, a maintenance burden, a potential attack surface, and an all around waste of time and space. There are rare cases where unused code is useful for API symmetry or to help the type checker or something, but this isn’t one of them. This code should just go away.

abhigyank commented:
> There are rare cases where unused code is useful for API symmetry.

Yeah I was going for this, so that a util could have a set and get function. Since, we don't have it as an API explicitly, we could let it go then. One can always re-add it later if required. We can keep the commit that removes it :)

andersk commented:
We don’t have a set function and a get function here. A get function should return exactly the data that you put into a set function—symmetry. This `getProxy` function, on the other hand, returns the result of some weird string-mangling logic and maybe also sets an environment variable as a side effect. Its only purpose was to support the specific library that we’re now removing.

manavmehta committed:
remove unused request-util

net.request handles proxy by itself and moving away from request module makes request-util unused.

manavmehta committed:
Remove unused getProxy() function from proxy-util

The function is now not used upon migration to electron's native net.request and thus is being removed

abhigyank commented:
@andersk Can you review this for merging?

andersk commented on app/main/linuxupdater.ts:
```diff
@@ -12,41 +10,48 @@ const logger = new Logger({
 });
 
-export function linuxUpdateNotification(): void {
+export function linuxUpdateNotification(Session: Electron.session): void {
```
Don’t capitalize the variable `Session`.

andersk commented on app/main/request.ts:
```diff
@@ -0,0 +1,149 @@
+		session: Session
+	};
+
+	return new Promise((resolve, reject) => {
```
Let’s try to extract the raw `Promise` logic into minimal shared helper functions and lean on `async`/`await` as much as possible. With a small fix to the Electron typings in `typings.d.ts` (justified by the [documentation](https://www.electronjs.org/docs/api/incoming-message#class-incomingmessage): “`IncomingMessage` implements the [Readable Stream](https://nodejs.org/api/stream.html#stream_readable_streams) interface”):

```typescript
declare module 'electron' {
    interface IncomingMessage extends NodeJS.ReadableStream {}
}
```

we can do this:

```typescript
import getStream from 'get-stream';

async function fetchResponse(request: ClientRequest): Promise<IncomingMessage> {
    return new Promise((resolve, reject) => {
        request.on("response", resolve);
        request.on("abort", reject);
        request.on("error", reject);
        request.end();
    });
}

export async function _getServerSettings(…): Promise<ServerConf> {
    const response = await fetchResponse(net.request({
        url: domain + '/api/v1/server_settings',
        session,
    }));
    if (response.statusCode !== 200) {
        throw response;  // maybe we want this to be an actual Error?
    }
    const {realm_name, realm_uri, realm_icon} = JSON.parse(await getStream.buffer(response));
    …
}
```

andersk commented on app/main/request.ts:
```diff
@@ -0,0 +1,149 @@
+	});
+};
+
+export const _saveServerIcon = async (Url: string, Session: Electron.session): Promise<string> => {
```
Don’t capitalize the variables `Url` and `Session`.

andersk commented on app/main/request.ts:
```diff
@@ -0,0 +1,149 @@
+	return `${dir}/${hash >>> 0}${extension}`;
+};
+
+export const _getServerSettings = async (domain: string, Session: Electron.session): Promise<ServerConf> => {
```
Don’t capitalize the variable `Session`.

andersk commented on app/main/request.ts:
```diff
@@ -0,0 +1,149 @@
+
+/* Request: reconnect-util */
+
+export const _isOnline = async (reconnectUtil: ReconnectUtil, Session: Electron.session): Promise<boolean> => {
```
Don’t capitalize the variable `Session`.

andersk commented on app/renderer/js/utils/reconnect-util.ts:
```diff
@@ -32,23 +30,8 @@ export default class ReconnectUtil {
-			}
-		});
+		const isOnlinePromise: Promise<boolean> = ipcRenderer.invoke('is-online', this);
+		return isOnlinePromise;
```
`return ipcRenderer.invoke('is-online', this)` seems fine.

andersk commented on app/main/request.ts:
```diff
@@ -0,0 +1,149 @@
+
+				response.on('end', () => {
+					resolve(filePath);
+				});
```
I think the `async` equivalent here would be
```typescript
const pipeline = util.promisify(stream.pipeline);
…
await pipeline(response, file);
return filePath;
```

andersk committed:
autoupdater: Use async/await.

Signed-off-by: Anders Kaseorg <anders@zulip.com>

andersk commented:
This also seems to trigger an extremely unfortunate bug where one of the renderer processes goes into a 100% CPU loop that eventually consumes all available memory. I’m trying to debug this…

andersk commented:
The bug had something to do with trying to serialize an entire `ReconnectUtil` over the IPC channel. We only need its `url`, so yeah, let’s not do that.

manavmehta commented:
Thanks @andersk. Though,

> one of the renderer processes goes into a 100% CPU loop that eventually consumes all available memory.

I didn't get this totally. Can you please explain it to me?

akashnimare commented:
Thanks, everyone 🤝
